File: js/reporters/text.ts
import pc from 'picocolors';
import type {
  ClassQualityReport,
  FileQualityReport,
  FunctionQualityReport,
  Grade,
  ProjectQualityReport,
  RefactoringFlag,
} from '../types.js';

export interface TextReporterOptions {
  verbose?: boolean;
  flaggedOnly?: boolean;
  scope?: 'function' | 'class' | 'file' | 'module';
}

// ─── Grade colors ─────────────────────────────────────────────────────────────

function gradeColor(grade: Grade, text: string): string {
  switch (grade) {
    case 'A':
      return pc.green(text);
    case 'B':
      return pc.cyan(text);
    case 'C':
      return pc.yellow(text);
    case 'D':
      return pc.red(text);
    case 'F':
      return pc.bgRed(pc.white(text));
  }
}

function scoreBar(score: number): string {
  const filled = Math.round(score / 10);
  const bar = '█'.repeat(filled) + '░'.repeat(10 - filled);
  return gradeColor(scoreToGrade(score), bar);
}

function scoreToGrade(score: number): Grade {
  if (score >= 80) return 'A';
  if (score >= 65) return 'B';
  if (score >= 50) return 'C';
  if (score >= 35) return 'D';
  return 'F';
}

// ─── Flag rendering ───────────────────────────────────────────────────────────

function renderFlag(flag: RefactoringFlag, indent = '    '): string {
  const sev = flag.severity === 'error' ? pc.red('[error]') : pc.yellow('[warn] ');
  return `${indent}${sev} ${flag.message}\n${indent}       ${pc.dim('→')} ${flag.suggestion}`;
}

function renderFlagList(flags: RefactoringFlag[]): string[] {
  if (flags.length === 0) return [];
  const lines: string[] = [];
  lines.push('    ' + pc.dim('Flags:'));
  for (const flag of flags) {
    lines.push(renderFlag(flag));
  }
  return lines;
}

// ─── Function row ─────────────────────────────────────────────────────────────

function renderFunctionHeader(fn: FunctionQualityReport): string {
  const icon = fn.needsRefactoring ? pc.red('✗') : pc.green('✓');
  const grade = gradeColor(fn.grade, `[${fn.grade}]`);
  const score = fn.needsRefactoring
    ? pc.red(`score: ${fn.score.toFixed(0)}`)
    : pc.green(`score: ${fn.score.toFixed(0)}`);

  let line = `  ${icon} ${pc.bold(fn.name)}()  ${grade}  ${score}`;

  if (fn.needsRefactoring) {
    line += pc.dim('  ← needs refactoring');
  }

  return line;
}

function renderVerboseMetrics(fn: FunctionQualityReport): string {
  const m = fn.metrics;
  return (
    `      CFC: ${m.cognitiveFlow.score} (${scoreToGrade(100 - m.cognitiveFlow.score * 4)})  ` +
    `DCI: ${m.dataComplexity.difficulty.toFixed(1)} (${scoreToGrade(100 - m.dataComplexity.difficulty)})  ` +
    `IRC: ${m.identifierReference.totalIrc.toFixed(1)}  ` +
    `Params: ${m.structural.parameterCount}  ` +
    `LOC: ${m.structural.loc}`
  );
}

function renderFunction(fn: FunctionQualityReport, opts: TextReporterOptions): string {
  const lines = [renderFunctionHeader(fn)];

  if (opts.verbose) {
    lines.push(renderVerboseMetrics(fn));
  }

  lines.push(...renderFlagList(fn.flags));

  return lines.join('\n');
}

// ─── File report ──────────────────────────────────────────────────────────────

function renderFileSummaryLine(report: FileQualityReport): string {
  const flagCount = report.flaggedFunctionCount;
  const total = report.functionCount;
  return (
    `File: ${gradeColor(report.grade, `${report.grade} — ${report.score.toFixed(1)}`)}` +
    (flagCount > 0
      ? pc.red(`  — ${flagCount} of ${total} function(s) need refactoring`)
      : pc.green(`  — all ${total} function(s) within thresholds`))
  );
}

function renderFileHeader(report: FileQualityReport): string[] {
  const gradeStr = gradeColor(report.grade, report.grade);
  return [
    '',
    pc.bold(`qualitas-ts: ${report.filePath}`),
    `${scoreBar(report.score)}  score: ${report.score.toFixed(1)}  grade: ${gradeStr}`,
    '',
  ];
}

function renderClassScopeSection(report: FileQualityReport, opts: TextReporterOptions): string[] {
  const lines: string[] = [];
  for (const cls of report.classes) {
    if (opts.flaggedOnly && !cls.needsRefactoring) continue;
    lines.push(
      `  ${pc.cyan(`class ${cls.name}`)}  ${gradeColor(cls.grade, cls.grade)}  score: ${cls.score.toFixed(1)}`,
    );
    lines.push(...renderFlagList(cls.flags));
  }
  return lines;
}

function renderFunctionScopeSection(
  report: FileQualityReport,
  opts: TextReporterOptions,
): string[] {
  const fns = opts.flaggedOnly
    ? report.functions.filter((f) => f.needsRefactoring)
    : report.functions;

  const lines = fns.map((fn) => renderFunction(fn, opts));

  for (const cls of report.classes) {
    lines.push(...renderClassMethods(cls, opts));
  }

  return lines;
}

function renderClassMethods(cls: ClassQualityReport, opts: TextReporterOptions): string[] {
  const methods = opts.flaggedOnly ? cls.methods.filter((m) => m.needsRefactoring) : cls.methods;

  if (opts.flaggedOnly && methods.length === 0) return [];

  return [
    '',
    `  ${pc.cyan(`class ${cls.name}`)}  ${gradeColor(cls.grade, cls.grade)}  score: ${cls.score.toFixed(1)}`,
    ...methods.map((m) => renderFunction(m, opts)),
  ];
}

export function renderFileReport(
  report: FileQualityReport,
  opts: TextReporterOptions = {},
): string {
  const lines = renderFileHeader(report);
  const scope = opts.scope ?? 'function';

  if (scope === 'file') {
    lines.push(renderFileSummaryLine(report));
    return lines.join('\n');
  }

  if (scope === 'class') {
    lines.push(...renderClassScopeSection(report, opts));
  } else {
    lines.push(...renderFunctionScopeSection(report, opts));
  }

  lines.push('');
  lines.push(renderFileSummaryLine(report));

  return lines.join('\n');
}

// ─── Project report ───────────────────────────────────────────────────────────

function renderProjectHeader(report: ProjectQualityReport): string[] {
  const s = report.summary;
  return [
    '',
    pc.bold(`qualitas-ts project: ${report.dirPath}`),
    `${scoreBar(report.score)}  score: ${report.score.toFixed(1)}  grade: ${gradeColor(report.grade, report.grade)}`,
    '',
    `  ${s.totalFiles} files  |  ${s.totalFunctions} functions  |  ` +
      pc.red(`${s.flaggedFunctions} need refactoring`),
    `  Grades: ${pc.green('A:' + s.gradeDistribution.a)}  ${pc.cyan('B:' + s.gradeDistribution.b)}  ` +
      `${pc.yellow('C:' + s.gradeDistribution.c)}  ${pc.red('D:' + s.gradeDistribution.d)}  ` +
      `${pc.bgRed(pc.white('F:' + s.gradeDistribution.f))}`,
  ];
}

function renderWorstFunctionsSection(worstFunctions: FunctionQualityReport[]): string[] {
  if (worstFunctions.length === 0) return [];
  const lines: string[] = ['', pc.bold('  Worst functions:')];
  for (const fn of worstFunctions.slice(0, 5)) {
    lines.push(
      `    ${pc.red('✗')} ${fn.location.file}  ${pc.bold(fn.name)}()  score: ${fn.score.toFixed(0)}  ${gradeColor(fn.grade, fn.grade)}`,
    );
  }
  return lines;
}

function renderProjectFileDetails(
  report: ProjectQualityReport,
  opts: TextReporterOptions,
): string[] {
  if (!opts.verbose && opts.flaggedOnly) return [];
  const lines: string[] = [''];
  for (const file of report.files) {
    if (opts.flaggedOnly && !file.needsRefactoring) continue;
    lines.push(renderFileReport(file, opts));
  }
  return lines;
}

export function renderProjectReport(
  report: ProjectQualityReport,
  opts: TextReporterOptions = {},
): string {
  const lines = renderProjectHeader(report);
  const scope = opts.scope ?? 'function';

  if (scope === 'module') {
    return lines.join('\n');
  }

  if (scope === 'function') {
    lines.push(...renderWorstFunctionsSection(report.worstFunctions));
  }

  lines.push(...renderProjectFileDetails(report, opts));

  return lines.join('\n');
}

File: crates/qualitas-cli/src/reporters/text.rs
use colored::Colorize;

use qualitas_core::scorer::thresholds::grade_from_score;
use qualitas_core::types::{
    ClassQualityReport, FileQualityReport, FunctionQualityReport, Grade, ProjectQualityReport,
    RefactoringFlag, Severity,
};

pub struct TextReporterOptions {
    pub verbose: bool,
    pub flagged_only: bool,
    pub scope: String,
}

impl Default for TextReporterOptions {
    fn default() -> Self {
        Self {
            verbose: false,
            flagged_only: false,
            scope: "function".to_string(),
        }
    }
}

// ─── Grade colors ─────────────────────────────────────────────────────────────

fn grade_color(grade: Grade, text: &str) -> String {
    match grade {
        Grade::A => text.green().to_string(),
        Grade::B => text.cyan().to_string(),
        Grade::C => text.yellow().to_string(),
        Grade::D => text.red().to_string(),
        Grade::F => text.white().on_red().to_string(),
    }
}

fn score_bar(score: f64) -> String {
    let filled = (score / 10.0).round() as usize;
    let filled = filled.min(10);
    let bar = format!(
        "{}{}",
        "\u{2588}".repeat(filled),
        "\u{2591}".repeat(10 - filled)
    );
    grade_color(grade_from_score(score, None), &bar)
}

// ─── Flag rendering ───────────────────────────────────────────────────────────

fn render_flag(flag: &RefactoringFlag, indent: &str) -> String {
    let sev = match flag.severity {
        Severity::Error => "[error]".red().to_string(),
        Severity::Warning => "[warn] ".yellow().to_string(),
        Severity::Info => "[info] ".dimmed().to_string(),
    };
    let arrow = "\u{2192}".dimmed().to_string();
    format!(
        "{indent}{sev} {}\n{indent}       {arrow} {}",
        flag.message, flag.suggestion
    )
}

// ─── Function row ─────────────────────────────────────────────────────────────

fn render_function(func: &FunctionQualityReport, opts: &TextReporterOptions) -> String {
    let icon = if func.needs_refactoring {
        "\u{2717}".red().to_string()
    } else {
        "\u{2713}".green().to_string()
    };
    let grade = grade_color(func.grade, &format!("[{}]", func.grade));
    let score_str = if func.needs_refactoring {
        format!("score: {}", func.score as u32).red().to_string()
    } else {
        format!("score: {}", func.score as u32).green().to_string()
    };

    let mut line = format!("  {icon} {}()  {grade}  {score_str}", func.name.bold());

    if func.needs_refactoring {
        line += &"  \u{2190} needs refactoring".dimmed().to_string();
    }

    let mut lines = vec![line];

    if opts.verbose {
        let m = &func.metrics;
        lines.push(format!(
            "      CFC: {} ({})  DCI: {:.1} ({})  IRC: {:.1}  Params: {}  LOC: {}",
            m.cognitive_flow.score,
            grade_from_score(100.0 - f64::from(m.cognitive_flow.score) * 4.0, None),
            m.data_complexity.difficulty,
            grade_from_score(100.0 - m.data_complexity.difficulty, None),
            m.identifier_reference.total_irc,
            m.structural.parameter_count,
            m.structural.loc,
        ));
    }

    if !func.flags.is_empty() {
        lines.push(format!("    {}", "Flags:".dimmed()));
        for flag in &func.flags {
            lines.push(render_flag(flag, "    "));
        }
    }

    lines.join("\n")
}

// ─── File report ──────────────────────────────────────────────────────────────

fn render_file_summary_line(report: &FileQualityReport) -> String {
    let flag_count = report.flagged_function_count;
    let total = report.function_count;
    let score_text = format!("{} \u{2014} {:.1}", report.grade, report.score);
    let grade_str = grade_color(report.grade, &score_text);

    let status = if flag_count > 0 {
        format!("  \u{2014} {flag_count} of {total} function(s) need refactoring")
            .red()
            .to_string()
    } else {
        format!("  \u{2014} all {total} function(s) within thresholds")
            .green()
            .to_string()
    };

    format!("File: {grade_str}{status}")
}

// ─── Extracted helper: render file header lines ───────────────────────────────

fn render_file_header(report: &FileQualityReport) -> Vec<String> {
    let grade_str = grade_color(report.grade, &report.grade.to_string());

    vec![
        String::new(),
        format!("qualitas: {}", report.file_path).bold().to_string(),
        format!(
            "{}  score: {:.1}  grade: {grade_str}",
            score_bar(report.score),
            report.score,
        ),
        String::new(),
    ]
}

// ─── Extracted helper: render flags for a class ───────────────────────────────

fn render_class_flags(flags: &[RefactoringFlag]) -> Vec<String> {
    let mut lines = Vec::new();
    if !flags.is_empty() {
        lines.push(format!("    {}", "Flags:".dimmed()));
        for flag in flags {
            lines.push(render_flag(flag, "    "));
        }
    }
    lines
}

// ─── Extracted helper: render class scope ─────────────────────────────────────

fn render_class_scope(report: &FileQualityReport, opts: &TextReporterOptions) -> Vec<String> {
    let mut lines = Vec::new();
    for cls in &report.classes {
        if opts.flagged_only && cls.methods.iter().all(|m| m.flags.is_empty()) {
            continue;
        }
        lines.push(format!(
            "  {}  {}  score: {:.1}",
            format!("class {}", cls.name).cyan(),
            grade_color(cls.grade, &cls.grade.to_string()),
            cls.score,
        ));
        lines.extend(render_class_flags(&cls.flags));
    }
    lines
}

// ─── Extracted helper: filter functions by flagged_only setting ────────────────

fn filter_functions(
    functions: &[FunctionQualityReport],
    flagged_only: bool,
) -> Vec<&FunctionQualityReport> {
    if flagged_only {
        functions.iter().filter(|f| !f.flags.is_empty()).collect()
    } else {
        functions.iter().collect()
    }
}

// ─── Extracted helper: render class with its methods in function scope ─────────

fn render_class_methods(cls: &ClassQualityReport, opts: &TextReporterOptions) -> Vec<String> {
    let class_fns = filter_functions(&cls.methods, opts.flagged_only);
    if opts.flagged_only && class_fns.is_empty() {
        return Vec::new();
    }

    let mut lines = vec![
        String::new(),
        format!(
            "  {}  {}  score: {:.1}",
            format!("class {}", cls.name).cyan(),
            grade_color(cls.grade, &cls.grade.to_string()),
            cls.score,
        ),
    ];
    for m in &class_fns {
        lines.push(render_function(m, opts));
    }
    lines
}

// ─── Extracted helper: render function scope ──────────────────────────────────

fn render_function_scope(report: &FileQualityReport, opts: &TextReporterOptions) -> Vec<String> {
    let mut lines = Vec::new();

    // Render file-scope before functions
    if let Some(fs) = &report.file_scope {
        if !opts.flagged_only || !fs.flags.is_empty() {
            lines.push(render_function(fs, opts));
        }
    }

    let fns = filter_functions(&report.functions, opts.flagged_only);
    for func in &fns {
        lines.push(render_function(func, opts));
    }

    for cls in &report.classes {
        lines.extend(render_class_methods(cls, opts));
    }

    lines
}

pub fn render_file_report(report: &FileQualityReport, opts: &TextReporterOptions) -> String {
    let mut lines = render_file_header(report);

    if opts.scope == "file" {
        lines.push(render_file_summary_line(report));
        return lines.join("\n");
    }

    if opts.scope == "class" {
        lines.extend(render_class_scope(report, opts));
    } else {
        // scope: 'function' (default) — full per-function detail
        lines.extend(render_function_scope(report, opts));
    }

    lines.push(String::new());
    lines.push(render_file_summary_line(report));

    lines.join("\n")
}

// ─── Extracted helper: render project header ──────────────────────────────────

fn render_project_header(report: &ProjectQualityReport) -> Vec<String> {
    let s = &report.summary;

    vec![
        String::new(),
        format!("qualitas project: {}", report.dir_path)
            .bold()
            .to_string(),
        format!(
            "{}  score: {:.1}  grade: {}",
            score_bar(report.score),
            report.score,
            grade_color(report.grade, &report.grade.to_string()),
        ),
        String::new(),
        format!(
            "  {} files  |  {} functions  |  {}",
            s.total_files,
            s.total_functions,
            if s.flagged_functions > 0 {
                format!("{} need refactoring", s.flagged_functions)
                    .red()
                    .to_string()
            } else {
                format!("{} need refactoring", s.flagged_functions)
                    .green()
                    .to_string()
            },
        ),
        format!(
            "  Grades: {}  {}  {}  {}  {}",
            format!("A:{}", s.grade_distribution.a).green(),
            format!("B:{}", s.grade_distribution.b).cyan(),
            format!("C:{}", s.grade_distribution.c).yellow(),
            format!("D:{}", s.grade_distribution.d).red(),
            format!("F:{}", s.grade_distribution.f).white().on_red(),
        ),
    ]
}

// ─── Extracted helper: render worst functions section ─────────────────────────

fn format_worst_fn_line(func: &FunctionQualityReport) -> String {
    let icon = if func.needs_refactoring {
        "\u{2717}".red().to_string()
    } else {
        "\u{2713}".green().to_string()
    };
    format!(
        "    {icon} {}  {}()  score: {}  {}",
        func.location.file,
        func.name.bold(),
        func.score as u32,
        grade_color(func.grade, &func.grade.to_string()),
    )
}

fn render_worst_functions_section(
    report: &ProjectQualityReport,
    flagged_only: bool,
) -> Vec<String> {
    let funcs: Vec<&FunctionQualityReport> = if flagged_only {
        report
            .worst_functions
            .iter()
            .filter(|f| !f.flags.is_empty())
            .take(10)
            .collect()
    } else {
        report.worst_functions.iter().take(5).collect()
    };

    if funcs.is_empty() {
        return vec![];
    }

    let header = if flagged_only {
        "  Functions with flags:"
    } else {
        "  Worst functions:"
    };
    let mut lines = vec![String::new(), header.bold().to_string()];
    for func in &funcs {
        lines.push(format_worst_fn_line(func));
    }
    lines
}

fn file_has_flags(file: &FileQualityReport) -> bool {
    file.file_scope
        .as_ref()
        .is_some_and(|fs| !fs.flags.is_empty())
        || file.functions.iter().any(|f| !f.flags.is_empty())
        || file
            .classes
            .iter()
            .any(|c| c.methods.iter().any(|m| !m.flags.is_empty()))
}

// ─── Extracted helper: render file details section ────────────────────────────

fn render_file_details_section(
    report: &ProjectQualityReport,
    opts: &TextReporterOptions,
) -> Vec<String> {
    let mut lines = vec![String::new()];
    for file in &report.files {
        if opts.flagged_only && !file_has_flags(file) {
            continue;
        }
        lines.push(render_file_report(file, opts));
    }
    lines
}

// ─── Project report ───────────────────────────────────────────────────────────

pub fn render_project_report(report: &ProjectQualityReport, opts: &TextReporterOptions) -> String {
    let mut lines = render_project_header(report);

    if opts.scope == "module" {
        return lines.join("\n");
    }

    if opts.scope == "function" {
        lines.extend(render_worst_functions_section(report, opts.flagged_only));
    }

    lines.extend(render_file_details_section(report, opts));

    lines.join("\n")
}
